File: src/csharp/Problems/MaxSubArray.cs
//https://leetcode.com/problems/maximum-subarray/

namespace LeetCode.Problems;

public sealed class MaxSubArray : ProblemBase
{
    [Theory]
    [ClassData(typeof(MaxSubArray))]
    public override void Test(object[] data) => base.Test(data);

    public override void AddTestCases()
        => Add(it => it.ParamArray("[-2,1,-3,4,-1,2,1,-5,4]").Result(6))
          .Add(it => it.ParamArray("[1]").Result(1))
          .Add(it => it.ParamArray("[-1]").Result(-1))
          .Add(it => it.ParamArray("[5,4,-1,7,8]").Result(23));

    private int Solution(int[] nums)
    {
        var sum = nums[0];
        var max = nums[0];
        for (var i = 1; i < nums.Length; i++)
        {
            sum = Math.Max(sum + nums[i], nums[i]);
            max = Math.Max(max, sum);
        }

        return max;
    }
}

File: src/go/problems/MaxSubArray.go
// https://leetcode.com/problems/two-sum/

package problems

func MaxSubArray(nums []int) int {
	sum := nums[0]
	max := nums[0]

	for i := 1; i < len(nums); i++ {
		sum = Max(sum+nums[i], nums[i])
		max = Max(max, sum)
	}

	return max
}

func Max(left int, right int) int {
	if left > right {
		return left
	}

	return right
}
